File: Session_06_DynamicProgramming_Classic/multi_language_implementations/longest_common_subsequence.java
import java.util.Scanner;

public class longest_common_subsequence {
    public static void solveLCS(String s1, String s2) {
        int m = s1.length();
        int n = s2.length();
        int[][] dp = new int[m + 1][n + 1];
        for (int i = 0; i <= m; i++) {
            for (int j = 0; j <= n; j++) {
                if (i == 0 || j == 0) {
                    dp[i][j] = 0;
                } else if (s1.charAt(i - 1) == s2.charAt(j - 1)) {
                    dp[i][j] = dp[i - 1][j - 1] + 1;
                } else {
                    dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
                }
            }
        }
        System.out.println("LCS Length: " + dp[m][n]);
        int index = dp[m][n];
        char[] lcsStr = new char[index];
        int i = m, j = n;
        while (i > 0 && j > 0) {
            if (s1.charAt(i - 1) == s2.charAt(j - 1)) {
                lcsStr[index - 1] = s1.charAt(i - 1);
                i--;
                j--;
                index--;
            } else if (dp[i - 1][j] > dp[i][j - 1]) {
                i--;
            } else {
                j--;
            }
        }
        System.out.println("LCS String: " + new String(lcsStr));
    }

    public static void main(String[] args) {
        String s1 = "STONE";
        String s2 = "LONEST";
        solveLCS(s1, s2);
        System.out.println();
        System.out.print("Press Enter to exit...");
        Scanner scanner = new Scanner(System.in);
        scanner.nextLine();
        scanner.close();
    }
}

File: Session_06_DynamicProgramming_Classic/multi_language_implementations/longest_common_subsequence.kt
import java.util.Scanner
import kotlin.math.max

fun solveLCS(s1: String, s2: String) {
    val m = s1.length
    val n = s2.length
    val dp = Array(m + 1) { IntArray(n + 1) }

    for (i in 0..m) {
        for (j in 0..n) {
            if (i == 0 || j == 0) {
                dp[i][j] = 0
            } else if (s1[i - 1] == s2[j - 1]) {
                dp[i][j] = dp[i - 1][j - 1] + 1
            } else {
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
            }
        }
    }
    println("LCS Length: ${dp[m][n]}")

    var index = dp[m][n]
    val lcsStr = CharArray(index)
    var i = m
    var j = n
    while (i > 0 && j > 0) {
        if (s1[i - 1] == s2[j - 1]) {
            lcsStr[index - 1] = s1[i - 1]
            i--
            j--
            index--
        } else if (dp[i - 1][j] > dp[i][j - 1]) {
            i--
        } else {
            j--
        }
    }
    println("LCS String: ${String(lcsStr)}")
}

fun main() {
    val s1 = "STONE"
    val s2 = "LONEST"
    solveLCS(s1, s2)
    println()
    print("Press Enter to exit...")
    Scanner(System.`in`).nextLine()
}
